File: host-windows/src/AppControl.Host/Security/EmojiTable.cs
namespace AppControl.Host.Security;

/// <summary>
/// Die 64 Symbole der SAS-Darstellung. 6 Bit pro Emoji, 5 Emojis = 30 Bit.
///
/// AUSWAHLKRITERIEN: visuell klar unterscheidbar, auf macOS und Windows aehnlich
/// dargestellt, in beiden Sprachen leicht benennbar (die Nutzer lesen sie sich am
/// Telefon vor). Keine Flaggen, keine Hautfarben-Modifier, keine Symbole, die
/// sich zwischen Plattformen stark unterscheiden.
///
/// DIE REIHENFOLGE IST TEIL DES PROTOKOLLS. Aendert sie sich, zeigen Host und
/// Viewer unterschiedliche Symbole fuer denselben Schluessel, und die Nutzer
/// brechen eine voellig gesunde Verbindung ab. Gegenstueck: EmojiTable.swift und
/// tools/crypto-vectors/generate.mjs - alle drei muessen identisch sein.
/// </summary>
public static class EmojiTable
{
    public static readonly string[] Symbols =
    [
        "\U0001F419", "\U0001F352", "\U0001F680", "\U0001F514", "\U0001F3A9", "\U0001F335", "\U0001F418", "\U0001F355",
        "\U00002693", "\U0001F3B8", "\U0001F98A", "\U0001F344", "\U0001F6B2", "\U0001F319", "\U0001F427", "\U0001F34B",
        "\U0001F3F0", "\U0001F3BA", "\U0001F98B", "\U0001F345", "\U0001F682", "\U0001F31F", "\U0001F422", "\U0001F347",
        "\U000026FA", "\U0001F3BB", "\U0001F981", "\U0001F349", "\U0001F6F5", "\U0001F308", "\U0001F433", "\U0001F35E",
        "\U0001F5FF", "\U0001F941", "\U0001F989", "\U0001F351", "\U0001F681", "\U00002600", "\U0001F41D", "\U0001F955",
        "\U0001F3D4", "\U0001F3B9", "\U0001F99C", "\U0001F369", "\U000026F5", "\U0001F33B", "\U0001F42C", "\U0001F9C0",
        "\U0001F3AA", "\U0001F3B7", "\U0001F9A9", "\U0001F353", "\U0001F69C", "\U00002744", "\U0001F43F", "\U0001F330",
        "\U0001F5FC", "\U0001FA95", "\U0001F994", "\U0001F34D", "\U0001F6F8", "\U0001F340", "\U0001F9AD", "\U0001F968",
    ];

    static EmojiTable()
    {
        // Ein Tippfehler in der Tabelle wuerde sich sonst erst beim Nutzer als
        // "die Emojis stimmen nicht ueberein" aeussern - dem am schwersten zu
        // diagnostizierenden Fehlerbild, das dieses Protokoll hat.
        if (Symbols.Length != 64)
            throw new InvalidOperationException($"SAS-Tabelle muss 64 Symbole haben, hat {Symbols.Length}");
        if (Symbols.Distinct().Count() != 64)
            throw new InvalidOperationException("SAS-Tabelle enthaelt Duplikate");
    }
}

File: viewer-macos/Sources/AppControlViewer/Security/EmojiTable.swift
import Foundation

/// Die 64 Symbole der SAS-Darstellung. 6 Bit pro Emoji, 5 Emojis = 30 Bit.
///
/// AUSWAHLKRITERIEN: visuell klar unterscheidbar, auf macOS und Windows ähnlich
/// dargestellt, in beiden Sprachen leicht benennbar (die Nutzer lesen sie sich am
/// Telefon vor). Keine Flaggen, keine Hautfarben-Modifier, keine Symbole, die sich
/// zwischen Plattformen stark unterscheiden.
///
/// DIE REIHENFOLGE IST TEIL DES PROTOKOLLS. Ändert sie sich, zeigen Host und Viewer
/// unterschiedliche Symbole für denselben Schlüssel, und die Nutzer brechen eine
/// völlig gesunde Verbindung ab. Gegenstück: EmojiTable.cs und
/// tools/crypto-vectors/generate.mjs — alle drei müssen identisch sein.
public enum EmojiTable {
    public static let symbols: [String] = [
    "\u{1F419}", "\u{1F352}", "\u{1F680}", "\u{1F514}", "\u{1F3A9}", "\u{1F335}", "\u{1F418}", "\u{1F355}",
    "\u{2693}", "\u{1F3B8}", "\u{1F98A}", "\u{1F344}", "\u{1F6B2}", "\u{1F319}", "\u{1F427}", "\u{1F34B}",
    "\u{1F3F0}", "\u{1F3BA}", "\u{1F98B}", "\u{1F345}", "\u{1F682}", "\u{1F31F}", "\u{1F422}", "\u{1F347}",
    "\u{26FA}", "\u{1F3BB}", "\u{1F981}", "\u{1F349}", "\u{1F6F5}", "\u{1F308}", "\u{1F433}", "\u{1F35E}",
    "\u{1F5FF}", "\u{1F941}", "\u{1F989}", "\u{1F351}", "\u{1F681}", "\u{2600}", "\u{1F41D}", "\u{1F955}",
    "\u{1F3D4}", "\u{1F3B9}", "\u{1F99C}", "\u{1F369}", "\u{26F5}", "\u{1F33B}", "\u{1F42C}", "\u{1F9C0}",
    "\u{1F3AA}", "\u{1F3B7}", "\u{1F9A9}", "\u{1F353}", "\u{1F69C}", "\u{2744}", "\u{1F43F}", "\u{1F330}",
    "\u{1F5FC}", "\u{1FA95}", "\u{1F994}", "\u{1F34D}", "\u{1F6F8}", "\u{1F340}", "\u{1F9AD}", "\u{1F968}",
    ]

    /// Wird beim Start einmal aufgerufen (siehe AppControlViewerApp.init).
    public static func validate() {
        precondition(symbols.count == 64, "SAS-Tabelle muss 64 Symbole haben")
        precondition(Set(symbols).count == 64, "SAS-Tabelle enthält Duplikate")
    }
}
